lib/OpenQA/Scheduler/Model/Jobs.pm

```diff
@@ -37,6 +37,82 @@ sub determine_scheduled_jobs ($self) {
     return $self->scheduled_jobs;
 }
 
+# reduces the matching workers of the $to_be_scheduled jobs to support pinning parallel jobs to single host
+sub _pick_slots_with_common_worker_host ($to_be_scheduled) {
+    # return early if we don't need to care about picking a common host for the given set of jobs
+    return undef unless any { $_->{parallel_pinning} } @$to_be_scheduled;
+    my $cluster_size = scalar @$to_be_scheduled;
+    return undef if $cluster_size < 2;
+
+    # pick slots on the first best worker host that has enough matching slots; let each job pick one slot per host
+    my %matching_worker_slots_by_host;
+    my %visited_worker_slots_by_id;
+    my $picked_matching_worker_slots = [];
+    for my $job (@$to_be_scheduled) {
+        # go though the list of matching worker slots and pick one slot per host
+        my %visited_worker_slots_by_host;
+        for my $matching_worker (@{$job->{matching_workers}}) {
+            my $id = $matching_worker->id;
+            next if exists $visited_worker_slots_by_id{$id};  # skip slots that have already been picked
+            my $host = $matching_worker->host;
+            next if $visited_worker_slots_by_host{$host}++;  # skip to pick only one slot per host
+            my $matching_worker_slots = ($matching_worker_slots_by_host{$host} //= []);
+            push @$matching_worker_slots, $matching_worker;
+            $visited_worker_slots_by_id{$id} = $job;
+            next if @$matching_worker_slots < $cluster_size;
+            $picked_matching_worker_slots = $matching_worker_slots;
+            last;
+        }
+
+        # go tough the list of matching workers again to re-visit hosts we could not pick a slot on
```

Comment:
Sounds like maybe a recursive call could help here?

Reply:
I was also thinking about making this recursive. In fact, to be able to solve this in all cases, it probably needed to be recursive. However, maybe it would be better to keep it simple (and thus also not increase the complexity of the algorithm making it scale even more badly¹). What I'm going to do for now is to extract redundant code to separate functions without introducing further recursions.

¹ I think right now it'll scale still well enough while also still be good enough. It doesn't try too badly to solve weird constellations of worker classes but is also not completely stupid about it.